functorch/functorch/csrc/BatchRulesLinearAlgebra.cpp

```diff
@@ -176,6 +182,181 @@ householder_product_batch_rule(const Tensor &input, c10::optional<int64_t> input
+  static inline Tensor check_and_reshape_input(const Tensor& tensor, optional<int64_t> batch_dim) {
+    TORCH_CHECK(tensor.dim() > 2, op_name, ": The input tensor A must have at least 2 dimensions.");
+    return moveBatchDimToFront(tensor, batch_dim);
+  }
+
+  static oneOutput apply_one(
+      const Tensor& tensor,
+      optional<int64_t> batch_dim,
+      T... extra_args) {
+    const auto tensor_ = check_and_reshape_input(tensor, batch_dim);
+    return std::make_tuple(Func(tensor_, std::forward<T>(extra_args)...), 0);
+  }
+
+  static twoOutputs apply_two(
+      const Tensor& tensor,
+      optional<int64_t> batch_dim,
+      T... extra_args) {
+    const auto tensor_ = check_and_reshape_input(tensor, batch_dim);
+    const auto res = Func(tensor_, std::forward<T>(extra_args)...);
+    return std::make_tuple(std::get<0>(res), 0, std::get<1>(res), 0);
+  }
+
+  static threeOutputs apply_three(
+      const Tensor& tensor,
+      optional<int64_t> batch_dim,
+      T... extra_args) {
+    const auto tensor_ = check_and_reshape_input(tensor, batch_dim);
+    const auto res = Func(tensor_, std::forward<T>(extra_args)...);
+    return std::make_tuple(std::get<0>(res), 0, std::get<1>(res), 0, std::get<2>(res), 0);
+  }
+
+  static fourOutputs apply_four(
+      const Tensor& tensor,
+      optional<int64_t> batch_dim,
+      T... extra_args) {
+    const auto tensor_ = check_and_reshape_input(tensor, batch_dim);
+    const auto res = Func(tensor_, std::forward<T>(extra_args)...);
+    return std::make_tuple(std::get<0>(res), 0, std::get<1>(res), 0, std::get<2>(res), 0, std::get<3>(res), 0);
+  }
+};
+
+template <char const *op_name, typename A, A a, typename C>
+struct LinalgCheckMatrixBinaryRuleHelper;
+
+template <char const *op_name, typename F, F Func, typename A, typename B, typename... T>
+struct LinalgCheckMatrixBinaryRuleHelper<op_name, F, Func, typelist<A, B, T...>> {
+  static twoOutputs apply_two(
+      const Tensor& first, optional<int64_t> first_bdim,
+      const Tensor& second, optional<int64_t> second_bdim,
+      T... extra_args) {
+    TORCH_CHECK(!first_bdim.has_value() || first.dim() > 2,
+              op_name, ": The input tensor A must have at least 2 dimensions.");
+    TORCH_CHECK(!second_bdim.has_value() || second.dim() > 2,
+              op_name, ": The input tensor A must have at least 2 dimensions.");
+    const auto tensor_other = _binary_pointwise_helper(first, first_bdim, second, second_bdim);
+    const auto tensor_ = std::get<0>(tensor_other);
+    const auto other_ = std::get<1>(tensor_other);
+    const auto res = Func(tensor_, other_, std::forward<T>(extra_args)...);
+    return std::make_tuple(std::get<0>(res), 0, std::get<1>(res), 0);
+  }
+};
+
+oneOutput cholesky_solve_batch_rule(
+    const Tensor& self, c10::optional<int64_t> self_bdim,
+    const Tensor& A, c10::optional<int64_t> A_bdim,
+    bool upper) {
+  TORCH_CHECK(!self_bdim.has_value() || self.dim() > 2,
```

Comment:
Let's use `rankWithoutBatchDim(self, self_bdim) >= 2` for clarity; ditto everywhere else